controllers/nstemplateset/cluster_resources.go

```diff
@@ -24,157 +19,131 @@ type clusterResourcesManager struct {
 		}
 	}
-	// go through all cluster resource kinds
-	for _, clusterResourceKind := range clusterResourceKinds {
-		gvkLogger := userTierLogger.WithValues("gvk", clusterResourceKind.gvk)
-		gvkCtx := log.IntoContext(ctx, gvkLogger)
 
-		gvkLogger.Info("ensuring cluster resources")
-		newObjs := make([]runtimeclient.Object, 0)
+	statusChanged := false
+	firstDeployment := nsTmplSet.Status.ClusterResources == nil || nsTmplSet.Status.ClusterResources.TemplateRef == ""
+	var failureStatusReason statusUpdater
+	if firstDeployment {
+		failureStatusReason = r.setStatusClusterResourcesProvisionFailed
+	} else {
+		failureStatusReason = r.setStatusUpdateFailed
+	}
 
-		// get all objects of the resource kind from the template (if the template is specified)
-		if tierTemplate != nil {
-			newObjs, err = tierTemplate.process(r.Scheme, map[string]string{
-				SpaceName: spacename,
-			}, retainObjectsOfSameGVK(clusterResourceKind.gvk))
+	changeStatusIfNeeded := func() error {
+		if !statusChanged {
+			if firstDeployment {
+				err = r.setStatusProvisioningIfNotUpdating(ctx, nsTmplSet)
+			} else {
+				err = r.setStatusUpdatingIfNotProvisioning(ctx, nsTmplSet)
+			}
 			if err != nil {
-				return false, r.wrapErrorWithStatusUpdateForClusterResourceFailure(gvkCtx, nsTmplSet, err,
-					"failed to process template for the cluster resources with the name '%s'", nsTmplSet.Spec.ClusterResources.TemplateRef)
+				return err
 			}
+			statusChanged = true
 		}
+		return nil
+	}
 
-		// list all existing objects of the cluster resource kind
-		currentObjects, err := clusterResourceKind.listExistingResourcesIfAvailable(ctx, r.Client, spacename, r.AvailableAPIGroups)
-		if err != nil {
-			return false, r.wrapErrorWithStatusUpdateForClusterResourceFailure(gvkCtx, nsTmplSet, err,
-				"failed to list existing cluster resources of GVK '%v'", clusterResourceKind.gvk)
+	for _, newObj := range newObjs {
+		if !shouldCreate(newObj, nsTmplSet) {
+			continue
 		}
 
-		// if there are more than one existing, then check if there is any that should be updated or deleted
-		if len(currentObjects) > 0 {
-			updatedOrDeleted, err := r.updateOrDeleteRedundant(gvkCtx, currentObjects, newObjs, tierTemplate, nsTmplSet)
-			if err != nil {
-				return false, r.wrapErrorWithStatusUpdate(gvkCtx, nsTmplSet, r.setStatusUpdateFailed,
-					err, "failed to update/delete existing cluster resources of GVK '%v'", clusterResourceKind.gvk)
-			}
-			if updatedOrDeleted {
-				return true, err
+		// by removing the new objects from the current objects, we are going to be left with the objects
+		// that should be removed from the cluster after we're done with this loop
+		for oldIdx, curObj := range currentObjects {
+			if curObj.GetName() == newObj.GetName() && curObj.GetNamespace() == newObj.GetNamespace() && curObj.GetObjectKind().GroupVersionKind() == newObj.GetObjectKind().GroupVersionKind() {
```

Comment:
there is already existing func `commonclient.SameGVKandName`

Reply:
Didn't know about this one. Thanks for spotting it 👍🏼